### DISCUSSION TITLE
Text update widget with "history"

### DISCUSSION THREAD

**Discussion by georgweiss** [ControlSystemStudio/phoebus#2305] [Ideas]
I have a request from a user:

A TextUpdate like widget capable of showing recent values, in a similar way as command line behavior of camonitor. This I guess would then be a text area with a scroll bar. Or maybe one could consider an extension to the TextUpdate where a "line count" property would turn it into a text area if line count > 1.

I've already pointed to alternatives like using a script that would print updates to the error log, or using the Python console. But maybe there are other options besides a new/extended widget?

**Comment by kasemir** [comment]
Looks like a very specific request that would be quite some work to implement and result in a widget that's only good for that one user.
Should the data be presented with time stamps? In what format? How many entries? Need to highlight a certain entry? Always auto-scroll to the most recent entry, or allow user to manually scroll up/down? Just display the last N values, or separately log them to a file to keep the longer history?

Existing alternatives:

Archive with data browser to display traces over time, and other archive-specific ways to query the data from scripts or web pages.
That's useful if the PV is of general interest and you need to keep the long-term history.

`camonitor ThatPVName` in a terminal. That's useful if just one person is interested in some PV for a little while, for example to debug a process.

If user cannot type `camonitor ThatPVName` in a terminal or doesn't like the font and color, create a custom python app that subscribes to the PV and presents the list in a custom GUI.

Alarm system. If the PV indicates significant problems, maybe it's a good alarm trigger, and the alarm history is then basically what you're looking for.

As soon as the widget goes beyond simply displaying a PV but "does" something, like logging the received values to a file, or correlating the data from several PVs to align them on a timeline I'd move that logic out of the display and into some service.

For example, an IOC can implement a 'bucket chain' type of records where you write to some 'input' record, which then copies the value to the next in a chain of say 10 records, with a parallel chain of records that indicate time stamp.
We use that on some devices to track the last N faults, with time stamp, and a highlight on that's reported as the 'first fault': 

<img width="396" alt="snap" src="https://user-images.githubusercontent.com/1932421/175296598-ec2b586f-8fc1-42a8-92d7-1e61fe43b957.png">

**Comment by georgweiss** [comment]
Not sure I agree with "...only good for that one user". We cannot know.
As for the features you list: a property would define the "buffer size", autoscroll=no, timestamps=probably.
Data browser would be tricky with string values.

As for your IOC example: how is it rendered? New-line separated string?

**Comment by kasemir** [comment]
For each row in that screenshot,  IOC has one record for the red dot indicator, the time stamp, and the message text.
In some cases there's a sequence on the IOC that checks things, then writes new { highlight?, message } to a pair of input records, which then set the time stamp and push the data into the bucket chain of records.
In another example, the logic is actually on a PLC, and the IOC just reads the N highlight, time stamp and message text values from the PLC.

Just us having several variants of that, implemented in PLC, IOC database and/or sequencer logic, suggests to me that it'll be hard to put all that into one widget with options to configure the details, but you're of course free to try.

**Comment by kasemir** [comment]
> New-line separated string?

With the phoebus text update widget, that would actually be a reasonable approach because you can set the widget to be 'interactive', so user can scroll, mark/copy selected text etc.
With EDM that's not possible, and since we started out with that and still need to support it, it's why we have one or more PVs per "line".
